File: Dsa/traversal.js
class Node{
    constructor(data){
        this.left=null
        this.data=data
        this.right=null
    }
}
class BinarySeacrh{
    constructor(root=null){
        this.root=root
    }

    preorder(Node=this.root){
       if(Node===null){
        return 
       }
       console.log(Node.data)
       this.preorder(Node.left)
       this.preorder(Node.right)
    }
    postOrder(Node=this.root){
        if(Node===null){
            return 
        }
        this.postOrder(Node.left)
        this.postOrder(Node.right)
        console.log(Node.data)
    }
    inOrder(Node=this.root){
        if(Node===null){
            return
        }
        this.inOrder(Node.left)
        console.log(Node.data)
        this.inOrder(Node.right)
    }
}


let search=new Node(34)
search.left=new Node(23)
search.right=new Node(56)
search.left.left=new Node(67)
search.left.right=new Node(34)
search.right.left=new Node(789)
search.right.right=new Node(567)
search.left.left.left=new Node(5678)
search.right.right.right=new Node(456778)


let tree=new BinarySeacrh(search)

//preorder implement
console.log("preOrder Traversal")
tree.preorder()

//postOrder implement
console.log("PostOrder Traversal")
tree.postOrder()

//inorder implement
console.log("Inorder Traversal")
tree.inOrder()

File: learnDsaSeries/python/traversal.py
class Node:
    def __init__(self,data):
        self.left=None
        self.right=None
        self.data=data

# perform the preorder traversal
def preorderTraversal(root):
    if root is not None:
        print(root.data,end=" ")
        preorderTraversal(root.left)
        preorderTraversal(root.right)

#perform the Inorder Traversal
def inorderTraversal(root):
    if root is not None:
        inorderTraversal(root.left)
        print(root.data,end=" ")
        inorderTraversal(root.right)
    

#perfom the postOrderTraversal
def postorderTraversal(root):
    if root is not None:
        postorderTraversal(root.left)
        postorderTraversal(root.right)
        print(root.data,end=" ")


#instance of class node 
root=Node(45)
root.left=Node(43)
root.left.left=Node(41)
root.left.left.left=Node(40)
root.right=Node(48)
root.right.right=Node(49)
root.right.right=Node(55)
root.right.right.left=Node(53)
root.right.left=Node(50)
root.left.right=Node(34)
root.left.left.right=Node(42)


print('Preorder Traverasal:')
preorderTraversal(root)
print("\n")---------
print('Inorder Traversal:')
inorderTraversal(root)
print("\n")
print('PostOrder Traversal')
postorderTraversal(root)
print("\n")


